### ISSUE TITLE
Create a Rendering Benchmark Sample Package

### ISSUE THREAD

**Issue by ilexp** [AdamsLair/duality#549]
### Summary

In order to keep track of performance and compare it across machines and Duality versions, it would be very nice to have a benchmark that can easily be installed and run.

### Analysis

- This came up originally in issue #326 and depends on some of the v3.0 rendering pipeline improvements.
- Use a render-to-texture approach.
  - Full control over multisampling and size, allowing to perform tests with a constant number of pixels to fill, vs. varying resolutions and settings per-user.
  - Allow users to change resolutions / multisampling / etc. via hotkey inside each benchmark.
- Benchmark scenes / cases:
  - Screen overlay vs world space
  - Lots of small (vertex count) renderers.
  - Few very big (vertex count) renderers.
- Use procedurally generated objects to keep the scene files small.

**Comment by ilexp** [comment]
### Progress

- Created a `benchmarks-wip` feature branch for working on this.
- Set up a benchmarks project.
- Set up a benchmarks sample package.

### Immediate ToDo

- Create a "baseline" testing scene with a simple sprite in the center of the screen and nothing else.
- Implement a special RenderSetup that renders everything to an internal offscreen target and then blits it to output, followed by drawing stats in screen space.
- Display minimalistic perf stats.
- Display instructions, similar to physics sample.
- Allow to adjust the setups internal offscreen target size and multisampling at runtime.
- Allow to switch between scenes via keyboard input, similar to physics sample.
- Implement various benchmarks.

**Comment by ilexp** [comment]
### Progress

- Created a `BenchmarkRenderSetup` to control rendering resolution, view size and antialiasing quality.
- Set up a baseline scene containing only a SpriteRenderer.

### Immediate ToDo

- Add diagnostics overlay to `BenchmarkRenderSetup`.
- Display minimalistic perf stats.
- Display instructions, similar to physics sample.
- Implement a sample controller Component that allows switching scenes and adjusting benchmark settings.
- Allow to adjust the setups internal offscreen target size and multisampling at runtime.
- Allow to switch between scenes via keyboard input, similar to physics sample.
- Implement various benchmarks.

**Comment by ilexp** [comment]
### Progress

- Added render setup info overlay.
- Added performance info overlay.
- Added benchmark info overlay.
- Created benchmark scenes for:
  - Large number of SpriteRenderers, alternating materials
  - Large number of SpriteRenderers, same material
  - Small number of "MultiSpriteRenderers"
- Removed the screen space sprite benchmarks from the ToDo for now, as they likely don't have a lot of additional value due to similar codepaths.

### Immediate ToDo

- Display GCs per minute instead of GCs overall.
- Implement benchmarks for:
  - Large number of "dummy renderers", which do not add any vertices at all, but otherwise behave as if they would, including positive BoundRadius, IsVisible calls, CalcZOffset calls and similar.
- Run the benchmark, take note about all the measurements and post them both here and in every rendering optimization related issue before starting.

**Comment by ilexp** [comment]
### Progress

- Measuring GCs per simulated minute (3600 frames) instead of the total count since startup.
- Added a dummy renderer sample where 10000 renderer objects are created, but none of them actually submit vertices to the device.

### Immediate ToDo

- Implement an automated test run that goes through all available benchmark scenes, waits a few seconds (~30? for average GC count), samples the displayed values and stores them in a text report.
- Run the benchmark, take note about all the measurements and post them both here and in every rendering optimization related issue before starting.

**Comment by ilexp** [comment]
### Progress

- Moved benchmark controller code to the core plugin, so it will remember things across scenes.
- Added an automated test run with text report generation.
- Merged back to `develop-3.0`.
